File: webapp/backend/pipeline.py
"""Pipeline orchestrator — runs the multi-agent analysis and emits SSE events."""

from __future__ import annotations

import asyncio
import sys
import time
import uuid
from pathlib import Path
from typing import AsyncGenerator

from .models import CompleteEvent, PipelineStage, StageEvent, StageStatus

# Add project root so we can import kalshi_client, fetchers, etc.
PROJECT_ROOT = Path(__file__).resolve().parent.parent.parent
if str(PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT))


class PipelineRun:
    """Tracks a single analysis run and streams SSE events."""

    def __init__(self, ticker: str):
        self.run_id = uuid.uuid4().hex[:12]
        self.ticker = ticker
        self.events: asyncio.Queue[StageEvent | CompleteEvent | None] = asyncio.Queue()
        self._task: asyncio.Task | None = None
        self._cancelled = False

    async def emit(self, event: StageEvent | CompleteEvent):
        await self.events.put(event)

    async def emit_stage(
        self, stage: PipelineStage, status: StageStatus, **kwargs
    ):
        await self.emit(StageEvent(stage=stage, status=status, **kwargs))

    def cancel(self):
        self._cancelled = True
        if self._task:
            self._task.cancel()

    async def run(self):
        """Execute the full pipeline, emitting events at each stage."""
        try:
            # --- Stage: Fetch ---
            await self.emit_stage(PipelineStage.FETCH, StageStatus.RUNNING)
            t0 = time.time()
            market_data = await self._fetch_market()
            await self.emit_stage(
                PipelineStage.FETCH, StageStatus.COMPLETE,
                duration_s=round(time.time() - t0, 1),
            )

            if self._cancelled:
                return

            # --- Stage: Evidence ---
            await self.emit_stage(PipelineStage.EVIDENCE, StageStatus.RUNNING)
            t0 = time.time()
            evidence_output = await self._run_evidence(market_data)
            await self.emit_stage(
                PipelineStage.EVIDENCE, StageStatus.COMPLETE,
                duration_s=round(time.time() - t0, 1),
            )

            if self._cancelled:
                return

            # --- Stage: Devil's Advocate ---
            await self.emit_stage(PipelineStage.DEVIL_ADVOCATE, StageStatus.RUNNING)
            t0 = time.time()
            da_output = await self._run_devils_advocate(market_data, evidence_output)
            await self.emit_stage(
                PipelineStage.DEVIL_ADVOCATE, StageStatus.COMPLETE,
                duration_s=round(time.time() - t0, 1),
            )

            if self._cancelled:
                return

            # --- Stages: Resolution + Chaos (parallel) ---
            await self.emit_stage(PipelineStage.RESOLUTION, StageStatus.RUNNING)
            await self.emit_stage(PipelineStage.CHAOS, StageStatus.RUNNING)
            t0 = time.time()

            resolution_output, chaos_output = await asyncio.gather(
                self._run_resolution(market_data, evidence_output, da_output),
                self._run_chaos(market_data, evidence_output, da_output),
            )

            elapsed = round(time.time() - t0, 1)
            await self.emit_stage(
                PipelineStage.RESOLUTION, StageStatus.COMPLETE, duration_s=elapsed,
            )
            await self.emit_stage(
                PipelineStage.CHAOS, StageStatus.COMPLETE, duration_s=elapsed,
            )

            if self._cancelled:
                return

            # --- Stage: Calibrator ---
            await self.emit_stage(PipelineStage.CALIBRATOR, StageStatus.RUNNING)
            t0 = time.time()
            calibrator_output = await self._run_calibrator(
                market_data, evidence_output, da_output,
                resolution_output, chaos_output,
            )
            await self.emit_stage(
                PipelineStage.CALIBRATOR, StageStatus.COMPLETE,
                duration_s=round(time.time() - t0, 1),
            )

            # --- Save report ---
            report_path = await self._save_report(
                market_data, evidence_output, da_output,
                resolution_output, chaos_output, calibrator_output,
            )

            await self.emit(CompleteEvent(run_id=self.run_id, report_path=report_path))

        except asyncio.CancelledError:
            pass
        except Exception as e:
            # Emit error on whatever stage was running
            await self.emit(StageEvent(
                stage=PipelineStage.FETCH,
                status=StageStatus.ERROR,
                detail=str(e),
            ))
        finally:
            await self.events.put(None)  # Signal stream end

    # --- Stage implementations (stubs for now — will integrate Agent SDK) ---

    async def _fetch_market(self) -> dict:
        """Fetch market data from Kalshi API."""
        import kalshi_client
        loop = asyncio.get_event_loop()
        return await loop.run_in_executor(None, kalshi_client.fetch_market, self.ticker)

    async def _run_evidence(self, market_data: dict) -> str:
        # TODO: integrate Claude Agent SDK
        await asyncio.sleep(0.5)  # placeholder
        return "[Evidence agent output placeholder]"

    async def _run_devils_advocate(self, market_data: dict, evidence: str) -> str:
        await asyncio.sleep(0.5)
        return "[Devil's Advocate output placeholder]"

    async def _run_resolution(self, market_data: dict, evidence: str, da: str) -> str:
        await asyncio.sleep(0.5)
        return "[Resolution agent output placeholder]"

    async def _run_chaos(self, market_data: dict, evidence: str, da: str) -> str:
        await asyncio.sleep(0.5)
        return "[Chaos agent output placeholder]"

    async def _run_calibrator(
        self, market_data: dict, evidence: str, da: str,
        resolution: str, chaos: str,
    ) -> str:
        await asyncio.sleep(0.5)
        return "[Calibrator output placeholder]"

    async def _save_report(
        self, market_data: dict, evidence: str, da: str,
        resolution: str, chaos: str, calibrator: str,
    ) -> str:
        """Save report to results/ directory and return the path."""
        from datetime import datetime

        results_dir = PROJECT_ROOT / "results"
        results_dir.mkdir(exist_ok=True)

        now = datetime.now()
        filename = f"{now.strftime('%Y-%m-%d_%H%M')}_{self.ticker}.md"
        path = results_dir / filename

        report = f"""# Analysis: {market_data.get('title', self.ticker)}
Generated: {now.isoformat()}
Ticker: {self.ticker}

## Calibrator Report
{calibrator}

## Evidence Agent
{evidence}

## Devil's Advocate
{da}

## Resolution Analysis
{resolution}

## Chaos Agent
{chaos}
"""
        path.write_text(report, encoding="utf-8")
        return str(path.relative_to(PROJECT_ROOT))


# Global registry of active runs
_active_runs: dict[str, PipelineRun] = {}


def get_run(run_id: str) -> PipelineRun | None:
    return _active_runs.get(run_id)


async def start_run(ticker: str) -> PipelineRun:
    run = PipelineRun(ticker)
    _active_runs[run.run_id] = run
    run._task = asyncio.create_task(run.run())
    return run


async def event_stream(run: PipelineRun) -> AsyncGenerator[str, None]:
    """Yield SSE-formatted strings from a pipeline run."""
    while True:
        event = await run.events.get()
        if event is None:
            break
        if isinstance(event, StageEvent):
            yield f"event: stage\ndata: {event.model_dump_json()}\n\n"
        elif isinstance(event, CompleteEvent):
            yield f"event: complete\ndata: {event.model_dump_json()}\n\n"

File: webapp/lib/pipeline.ts
import { spawn } from "child_process";
import fs from "fs";
import path from "path";
import { fetchMarket } from "./kalshi";
import type {
  MarketData,
  ParsedMarket,
  EventData,
  PipelineEvent,
  PipelineStage,
  StageEvent,
  ProgressEvent,
  CompleteEvent,
  ToolCategory,
} from "./types";
import { isEventData } from "./types";
import * as prompts from "./prompts";

// --- Claude CLI runner ---

/**
 * Run a Claude Code agent as a subprocess using the user's Max subscription.
 * Uses `claude -p --model <model> --output-format stream-json` for real-time progress.
 * Calls onProgress with human-readable status updates as the agent works.
 */
interface ToolProgress {
  detail: string;
  toolName: string;
  toolCategory: ToolCategory;
}

function classifyTool(toolName: string, input: Record<string, unknown>): ToolProgress {
  if (toolName === "WebSearch" || toolName === "web_search") {
    return {
      detail: `Searching: ${input.query || "..."}`,
      toolName: "WebSearch",
      toolCategory: "search",
    };
  }
  if (toolName === "Bash") {
    const cmd = (input.command as string) || "";
    if (cmd.includes("cli.ts cross-market")) {
      const q = cmd.match(/--query\s+"([^"]+)"/)?.[1] || "";
      return { detail: `Cross-market lookup: ${q}`, toolName: "cross-market", toolCategory: "fetcher" };
    }
    if (cmd.includes("cli.ts whitehouse")) {
      const q = cmd.match(/--search\s+"([^"]+)"/)?.[1] || "";
      return { detail: `White House search: ${q}`, toolName: "whitehouse", toolCategory: "fetcher" };
    }
    if (cmd.includes("cli.ts oira")) {
      const q = cmd.match(/--search\s+"([^"]+)"/)?.[1] || "";
      return { detail: `OIRA/Fed Register: ${q}`, toolName: "oira", toolCategory: "fetcher" };
    }
    if (cmd.includes("cli.ts fec")) {
      const who = cmd.match(/--(?:candidate|committee)\s+"([^"]+)"/)?.[1] || "";
      return { detail: `FEC lookup: ${who}`, toolName: "fec", toolCategory: "fetcher" };
    }
    if (cmd.includes("cli.ts polling")) {
      const race = cmd.match(/--race\s+"([^"]+)"/)?.[1] || "";
      return { detail: `Polling data: ${race}`, toolName: "polling", toolCategory: "fetcher" };
    }
    return { detail: `Running: ${cmd.slice(0, 80)}`, toolName: "Bash", toolCategory: "bash" };
  }
  return { detail: `Using ${toolName}...`, toolName, toolCategory: "thinking" };
}

async function runAgent(
  model: "haiku" | "sonnet" | "opus",
  prompt: string,
  allowedTools: string[] = ["WebSearch", "Bash"],
  onProgress?: (progress: ToolProgress) => void,
  signal?: AbortSignal,
): Promise<string> {
  return new Promise((resolve, reject) => {
    const args = [
      "-p",
      "--model", model,
      "--output-format", "stream-json",
      "--verbose",
      "--no-session-persistence",
      "--dangerously-skip-permissions",
    ];

    if (allowedTools.length > 0) {
      args.push("--allowedTools", ...allowedTools);
    } else {
      args.push("--disallowedTools", "Bash", "WebSearch", "Read", "Edit", "Write", "Glob", "Grep");
    }

    const env = { ...process.env };
    delete env.CLAUDECODE;

    const proc = spawn("claude", args, {
      stdio: ["pipe", "pipe", "pipe"],
      cwd: path.resolve(process.cwd(), ".."),
      env,
      shell: true,
    });

    let buffer = "";
    let resultText = "";
    let stderr = "";

    proc.stdout.on("data", (chunk: Buffer) => {
      buffer += chunk.toString();

      // Parse newline-delimited JSON
      const lines = buffer.split("\n");
      buffer = lines.pop() || ""; // keep incomplete line in buffer

      for (const line of lines) {
        if (!line.trim()) continue;
        try {
          const event = JSON.parse(line);
          // Extract progress info from streaming events
          if (event.type === "assistant" && event.message?.content) {
            for (const block of event.message.content) {
              if (block.type === "tool_use" && onProgress) {
                const toolName = block.name || "tool";
                const input = block.input || {};
                onProgress(classifyTool(toolName, input));
              }
            }
          }
          // Capture the final result
          if (event.type === "result") {
            resultText = event.result || "";
          }
        } catch {
          // Skip malformed JSON lines
        }
      }
    });

    proc.stderr.on("data", (chunk: Buffer) => {
      stderr += chunk.toString();
    });

    proc.on("close", (code) => {
      if (code === 0 && resultText) {
        resolve(resultText);
      } else if (code === 0) {
        // Fallback: no result event found, shouldn't happen
        resolve(buffer.trim());
      } else {
        reject(new Error(`Claude agent exited with code ${code}: ${stderr.slice(0, 500)}`));
      }
    });

    proc.on("error", (err) => {
      reject(new Error(`Failed to spawn claude: ${err.message}`));
    });

    // Kill subprocess if cancelled
    if (signal) {
      if (signal.aborted) {
        proc.kill();
        reject(new Error("Cancelled"));
        return;
      }
      signal.addEventListener("abort", () => {
        proc.kill();
      });
    }

    proc.stdin.write(prompt);
    proc.stdin.end();
  });
}

// --- Pipeline run ---

export class PipelineRun {
  runId: string;
  ticker: string;
  events: PipelineEvent[] = [];
  reportContent: string | null = null;
  private listeners: ((event: PipelineEvent | null) => void)[] = [];
  private abortController = new AbortController();
  private running = false;

  constructor(ticker: string) {
    this.runId = Math.random().toString(36).slice(2, 14);
    this.ticker = ticker;
  }

  get cancelled() {
    return this.abortController.signal.aborted;
  }

  private emit(event: PipelineEvent) {
    this.events.push(event);
    for (const listener of this.listeners) listener(event);
  }

  private emitStage(
    stage: StageEvent["stage"],
    status: StageEvent["status"],
    extra?: Partial<StageEvent>,
  ) {
    this.emit({ kind: "stage", stage, status, ...extra });
  }

  private emitProgress(stage: PipelineStage, progress: ToolProgress) {
    this.emit({
      kind: "progress",
      stage,
      detail: progress.detail,
      toolName: progress.toolName,
      toolCategory: progress.toolCategory,
      timestamp: Date.now(),
    } as ProgressEvent);
  }

  private progressCallback(stage: PipelineStage) {
    return (progress: ToolProgress) => this.emitProgress(stage, progress);
  }

  subscribe(fn: (event: PipelineEvent | null) => void): () => void {
    this.listeners.push(fn);
    return () => {
      this.listeners = this.listeners.filter((l) => l !== fn);
    };
  }

  cancel() {
    this.abortController.abort();
  }

  isRunning() {
    return this.running;
  }

  private signalEnd() {
    for (const listener of this.listeners) listener(null);
  }

  async run(): Promise<void> {
    this.running = true;

    try {
      // --- Fetch ---
      this.emitStage("fetch", "running");
      const t0 = Date.now();
      const marketData = await fetchMarket(this.ticker);
      this.emitStage("fetch", "complete", {
        durationS: (Date.now() - t0) / 1000,
      });

      if (this.cancelled) return;

      const isEvent = isEventData(marketData);

      // Build context strings
      let title: string;
      let resolutionCriteria: string;
      let closeDate: string;
      let yesPrice: string;
      let volume: number;
      let outcomesText = "";
      let subText = "";

      if (isEvent) {
        const ed = marketData as EventData;
        title = ed.title;
        resolutionCriteria = ed.markets[0]?.resolutionCriteria || "";
        closeDate = ed.markets[0]?.closeDate || "";
        yesPrice = "";
        volume = ed.markets.reduce((s, m) => s + m.volume, 0);
        outcomesText = ed.markets
          .map(
            (m, i) =>
              `${i + 1}. ${m.yesSubTitle || m.title} — YES ${(m.yesPrice * 100).toFixed(0)}%`,
          )
          .join("\n");
        subText =
          ed.subThresholdMarkets.length > 0
            ? ed.subThresholdMarkets
                .map(
                  (m) =>
                    `- ${m.yesSubTitle || m.title} — YES ${(m.yesPrice * 100).toFixed(1)}%`,
                )
                .join("\n")
            : "None";
      } else {
        const pm = marketData as ParsedMarket;
        title = pm.title;
        resolutionCriteria = pm.resolutionCriteria;
        closeDate = pm.closeDate;
        yesPrice = `${(pm.yesPrice * 100).toFixed(0)}%`;
        volume = pm.volume;
      }

      // --- Evidence ---
      this.emitStage("evidence", "running");
      const t1 = Date.now();
      const evidencePrompt = isEvent
        ? prompts.evidenceEvent(title, closeDate, resolutionCriteria, outcomesText)
        : prompts.evidenceBinary(title, resolutionCriteria, closeDate, yesPrice);

      const signal = this.abortController.signal;
      const evidenceOutput = await runAgent("haiku", evidencePrompt, ["WebSearch", "Bash"], this.progressCallback("evidence"), signal);
      this.emitStage("evidence", "complete", {
        durationS: (Date.now() - t1) / 1000,
      });

      if (this.cancelled) return;

      // Extract sources pool
      const sourcesPool = extractSourcesPool(evidenceOutput);

      // --- Devil's Advocate ---
      this.emitStage("devil_advocate", "running");
      const t2 = Date.now();
      const daPrompt = isEvent
        ? prompts.devilsAdvocateEvent(title, closeDate, outcomesText, evidenceOutput, sourcesPool)
        : prompts.devilsAdvocateBinary(title, resolutionCriteria, closeDate, yesPrice, evidenceOutput, sourcesPool);

      const daOutput = await runAgent("haiku", daPrompt, ["WebSearch", "Bash"], this.progressCallback("devil_advocate"), signal);
      this.emitStage("devil_advocate", "complete", {
        durationS: (Date.now() - t2) / 1000,
      });

      if (this.cancelled) return;

      // --- Resolution + Chaos (parallel) ---
      this.emitStage("resolution", "running");
      this.emitStage("chaos", "running");
      const t3 = Date.now();

      const resolutionPrompt = isEvent
        ? prompts.resolutionEvent(title, closeDate, resolutionCriteria, outcomesText, evidenceOutput, daOutput)
        : prompts.resolutionBinary(title, resolutionCriteria, closeDate, evidenceOutput, daOutput);

      const chaosPrompt = isEvent
        ? prompts.chaosEvent(title, closeDate, outcomesText, subText, evidenceOutput, daOutput)
        : prompts.chaosBinary(title, resolutionCriteria, closeDate, yesPrice, evidenceOutput, daOutput);

      const [resolutionOutput, chaosOutput] = await Promise.all([
        runAgent("sonnet", resolutionPrompt, ["WebSearch"], this.progressCallback("resolution"), signal),
        runAgent("haiku", chaosPrompt, ["WebSearch"], this.progressCallback("chaos"), signal),
      ]);

      const elapsed3 = (Date.now() - t3) / 1000;
      this.emitStage("resolution", "complete", { durationS: elapsed3 });
      this.emitStage("chaos", "complete", { durationS: elapsed3 });

      if (this.cancelled) return;

      // --- Calibrator ---
      this.emitStage("calibrator", "running");
      const t4 = Date.now();
      const calibratorPrompt = isEvent
        ? prompts.calibratorEvent(title, closeDate, outcomesText, subText, volume, evidenceOutput, daOutput, resolutionOutput, chaosOutput)
        : prompts.calibratorBinary(title, resolutionCriteria, closeDate, yesPrice, volume, evidenceOutput, daOutput, resolutionOutput, chaosOutput);

      const calibratorOutput = await runAgent("sonnet", calibratorPrompt, [], this.progressCallback("calibrator"), signal);
      this.emitStage("calibrator", "complete", {
        durationS: (Date.now() - t4) / 1000,
      });

      // --- Save report ---
      const reportPath = saveReport(
        this.ticker,
        title,
        calibratorOutput,
        evidenceOutput,
        daOutput,
        resolutionOutput,
        chaosOutput,
      );
      this.reportContent = fs.readFileSync(
        path.resolve(process.cwd(), "..", reportPath),
        "utf-8",
      );

      this.emit({
        kind: "complete",
        runId: this.runId,
        reportPath,
      } as CompleteEvent);
    } catch (err) {
      if (this.cancelled) {
        // Don't emit error for intentional cancellation
        return;
      }
      // Find the running stage and mark it as error
      const runningStage = (["fetch", "evidence", "devil_advocate", "resolution", "chaos", "calibrator"] as PipelineStage[])
        .find((s) => {
          const ev = this.events.findLast((e) => e.kind === "stage" && (e as StageEvent).stage === s);
          return ev && (ev as StageEvent).status === "running";
        });
      this.emitStage(runningStage || "fetch", "error", {
        detail: err instanceof Error ? err.message : String(err),
      });
    } finally {
      this.running = false;
      this.signalEnd();
    }
  }
}

function extractSourcesPool(text: string): string {
  const marker = "## SOURCES POOL";
  const idx = text.indexOf(marker);
  if (idx === -1) return "";
  return text.slice(idx);
}

function saveReport(
  ticker: string,
  title: string,
  calibrator: string,
  evidence: string,
  da: string,
  resolution: string,
  chaos: string,
): string {
  const projectRoot = path.resolve(process.cwd(), "..");
  const resultsDir = path.join(projectRoot, "results");
  if (!fs.existsSync(resultsDir)) fs.mkdirSync(resultsDir, { recursive: true });

  const now = new Date();
  const dateStr = now.toISOString().slice(0, 10);
  const timeStr = `${String(now.getHours()).padStart(2, "0")}${String(now.getMinutes()).padStart(2, "0")}`;
  const filename = `${dateStr}_${timeStr}_${ticker}.md`;
  const filePath = path.join(resultsDir, filename);

  const report = `# Analysis: ${title}
Generated: ${now.toISOString()}
Ticker: ${ticker}

## Calibrator Report
${calibrator}

## Evidence Agent
${evidence}

## Devil's Advocate
${da}

## Resolution Analysis
${resolution}

## Chaos Agent
${chaos}
`;

  fs.writeFileSync(filePath, report, "utf-8");
  return `results/${filename}`;
}

// --- Global run registry ---
// Attach to globalThis so the map survives Next.js HMR in dev mode

const globalForRuns = globalThis as unknown as { __pipelineRuns?: Map<string, PipelineRun> };
if (!globalForRuns.__pipelineRuns) {
  globalForRuns.__pipelineRuns = new Map();
}
const activeRuns = globalForRuns.__pipelineRuns;

export function getRun(runId: string): PipelineRun | undefined {
  return activeRuns.get(runId);
}

export function startRun(ticker: string): PipelineRun {
  const run = new PipelineRun(ticker);
  activeRuns.set(run.runId, run);
  run.run();
  return run;
}
